dhall-csv/src/Dhall/Csv.hs

```diff
@@ -1,4 +1,89 @@
-module Dhall.Csv (hello) where
+{-#LANGUAGE OverloadedStrings#-}
 
-hello :: String
-hello = "Hello Dhall to CSV!"
+module Dhall.Csv (
+      dhallToCsv
+    , codeToValue
+    ) where
+
+import Control.Exception            (Exception, throwIO)
+import Data.Csv                     (ToField (..))
+import Data.Maybe                   (fromMaybe)
+import Data.Sequence                (Seq)
+import Data.Text                    (Text)
+import Data.Text.Prettyprint.Doc    (Pretty)
+import Data.Void                    (Void)
+import Dhall.Core                   (Expr, DhallDouble (..))
+import Dhall.Import                 (SemanticCacheMode (..))
+import Dhall.Util                   (_ERROR)
+
+import qualified Data.Csv
+import qualified Data.Foldable
+import qualified Data.Text
+import qualified Data.Text.Prettyprint.Doc.Render.Text as Pretty
+import qualified Dhall.Core                            as Core
+import qualified Dhall.Import
+import qualified Dhall.Map
+import qualified Dhall.Parser
+import qualified Dhall.Pretty
+import qualified Dhall.TypeCheck
+import qualified Dhall.Util
+import qualified System.FilePath
+
+data CompileError = Unsupported (Expr Void Void)
+
+instance Show CompileError where
+    show (Unsupported e) =
+        Data.Text.unpack $
+            _ERROR <> ": Cannot translate to CSV                                             \n\
+            \                                                                                \n\
+            \Explanation: Only records of primitive values can be                            \n\
+            \translated from Dhall to CSV.                                                   \n\
+            \                                                                                \n\
+            \The following Dhall expression could not be translated to CSV:                  \n\
+            \                                                                                \n\
+            \" <> insert e
+
+instance Exception CompileError
+
+insert :: Pretty a => a -> Text
+insert = Pretty.renderStrict . Dhall.Pretty.layout . Dhall.Util.insert
+
+dhallToCsv
+    :: Expr s Void
+    -> Either CompileError (Seq Data.Csv.NamedRecord)
+dhallToCsv e0 = listConvert $ Core.alphaNormalize $ Core.normalize e0
```

Comment:
I think alpha-normalization doesn't really help us with the conversion to CSV. It might make some errors less useful though.

```suggestion
dhallToCsv e0 = listConvert $ Core.normalize e0
```

Reply:
I'm not quite sure of what alpha-normalization does, could you give a quick explanation? I put `alphaNormalize` there because it was in `dhall-json` package and I thought it was important 😅 but I'd like to know what it does and when I should use it

Reply:
α-normalization is used to "canonicalize" expressions that contain bound variables. For example, you can use it to check that lambdas with different variable names are equivalent:

```
$ echo '\(x : Natural) -> x + 1' | dhall --alpha
\(_ : Natural) -> _ + 1
$ echo '\(y : Natural) -> y + 1' | dhall --alpha
\(_ : Natural) -> _ + 1
```

I'm not sure why `dhall-json` uses it – I suspect it's unnecessary, but @Gabriel439 may know better.

I think we can do without it for now. If it turns out that we need it, we can easily add it.

Reply:
The reason `dhall-json` uses it is because it has to pattern match on functions that have a certain type (specifically functions of type `Prelude.JSON.Type`, and canonicalizing the variable names makes this logic much simpler

`dhall-csv` definitely doesn't need it because it doesn't have to convert functions

Reply:
Thanks!

Reply:
You're welcome! 🙂